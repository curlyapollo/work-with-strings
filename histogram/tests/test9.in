#pragma once

#include <string>


class Animal {
public:
    virtual std::string voice() const {
        return "Not implemented yet";
    }
    virtual ~Animal() {
    }
};


class Tiger: public Animal {
    std::string voice() const override {
        return "Rrrr";
    }
};


class Wolf: public Animal {
    std::string voice() const override {
        return "Wooo";
    }
};


class Fox: public Animal {
    std::string voice() const override {
        return "Tyaf";
    }
};

#include <iostream>
#include "animals.h"
#include <stdexcept>
#include <vector>
#include <memory>

using Zoo = std::vector<std::unique_ptr<Animal>>;

Zoo CreateZoo() {
    Zoo zoo;
    std::string word;
    while (std::cin >> word) {
        if (word == "Tiger") {
            zoo.push_back(std::unique_ptr<Tiger>(new Tiger()));
        } else if (word == "Wolf") {
            zoo.push_back(std::unique_ptr<Wolf>(new Wolf()));
        } else if (word == "Fox") {
            zoo.push_back(std::unique_ptr<Fox>(new Fox()));
        } else
            throw std::runtime_error("Unknown animal!");
    }
    return zoo;
}

void Process(const Zoo& zoo) {
    for (const auto& animal : zoo) {
        std::cout << animal->voice() << "\n";
    }
}


int main() {
    auto l = CreateZoo();
    Process(l);
    return 0;
}
