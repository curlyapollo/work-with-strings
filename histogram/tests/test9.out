if 3
else 3
for 1
int 1
void 1
